Enable HTTP persistence when using Session with Invoke-WebRequest and Invoke-RestMethod

stevenebutler opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

This PR addresses [12764](https://github.com/PowerShell/PowerShell/issues/12764) by using persistent HTTP connections when Invoke-WebRequest uses the -Session variable. It does this by 
* tracking the  HttpClient object in the saved WebRequestSession
* monitoring properties in WebRequestSession and rebuilding the HttpClient if they are changed. This is done to ensure that existing scripts that might tweak parameters between invocations of IWR actually apply the changes so they will behave in a backwards compatible manner. When anything that impacts the HttpClient is changed, the persistent connection is abandoned by disposing the HttpClient and underlying HttpClientHandler and constructing a new one. A warning is logged when this occurs as it has a performance impact (though no worse than the existing behaviour).
* To support the point 2 above, many switches were added as internal set-only properties to the WebRequestSession so changes between invocations can be detected and the handler rebuilt appropriately.

## PR Context

This PR makes a considerable difference to performance when using Invoke-WebRequest or Invoke-RestMethod with sessions when multiple requests are made to the same host. This is particularly evident for HTTPS requests, as each HTTPS connection can take hundreds of milliseconds to establish.

This PR as written passes existing tests in `WebCmdlets.Tests.ps1` and additional WebSession tests that verify the HttpClient is rebuilt when appropriate. This PR is submitted as WIP for feedback following advice on https://github.com/PowerShell/PowerShell/issues/12764 

The PR is attempting to provide as much backwards compatibility as possible at the expense of the complexity of tracking how the HttpClient used in the WebSession was constructed and recreating the HttpClient in the case it would have been constructed differently between invocations. Some feedback on whether this is a reasonable approach would help shape the PR.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

stevenebutler committed:
Enable HTTP persistence when using Session

stevenebutler commented:
@microsoft-github-policy-service agree

stevenebutler committed:
Address codefactor issues

stevenebutler committed:
Merge branch 'PowerShell:master' into 12764-support-persistent-http-connections-in-cmdlets

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -105,8 +259,16 @@ public WebRequestSession()
+        /// <summary>
+        /// Gets a value indicating whether a new handler must be created.
+        /// </summary>
+        internal bool Changed { get => _changed; }
```
```suggestion
        internal bool Changed => _changed; 
```

stevenebutler replied:
Updated as suggested.

CarloToso commented:
Maybe you could do something like this for WebRequestSession.cs (it's just an example I think it can be done better)

```cs
        internal int MaxAutomaticRedirections
        {
            get => _maxAutomaticRedirections;
            set => SetVariable(typeof(int), _maxAutomaticRedirections, value);
        }

        internal dynamic SetVariable(Type type, object oldValue, object newValue)
        {
            dynamic convertedOldValue = Convert.ChangeType(oldValue, type);
            dynamic convertedNewValue = Convert.ChangeType(newValue, type);
            if (convertedOldValue != convertedNewValue)
            {
                _changed = true;
                return convertedNewValue;
            }
            else
            {
                return convertedOldValue;
            }
        }
```

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -894,6 +903,10 @@ internal virtual void PrepareSession()
                 // Store the UserAgent string
                 WebSession.UserAgent = UserAgent;
             }
+            if (MyInvocation.BoundParameters.ContainsKey("NoProxy"))
```
```suggestion
            if (NoProxy.IsPresent)
```

CarloToso replied:
Maybe you could even remove the if and leave `WebSession.NoProxy = NoProxy.IsPresent;`

stevenebutler replied:
This one is tricky since it's a switch, the -NoProxy flag will have been given when creating the WebSession, but if it's not present on subsequent requests do we assume they just don't feel the need to repeat the -NoProxy because they're using a session, or do we assume they no longer want the proxy. 

I was using presence of the switch to detect that but on reflection I think it's still not going to do that right. Perhaps proxy related settings should just be ignored on subsequent requests when a websession has been requested?

If we identified some of the items that we won't support changing on the fly in a session we could potentially simplify the amount of state being monitored for change.

stevenebutler replied:
Without that presence check, this script will cause a reconnect on the first loop through using the websession (similarly for NoProxy and SslProtocol args):

```powershell
Measure-Command {
     Write-Host (Invoke-WebRequest https://postman-echo.com/get?i=1 -SessionVariable Session -SslProtocol Tls12 -NoProxy).StatusCode
    1..10 | ForEach-Object {
         write-host (Invoke-WebRequest https://postman-echo.com/get?i=$_ -WebSession $Session).StatusCode
     }
 }
```
With the check in, the only way to force it to see the change to the -NoProxy flag is to provide it explicitly set to $false

```powershell
Measure-Command {
     Write-Host (Invoke-WebRequest https://postman-echo.com/get?i=1 -SessionVariable Session -SslProtocol Tls12 -NoProxy).StatusCode
    1..10 | ForEach-Object {
         write-host (Invoke-WebRequest https://postman-echo.com/get?i=$_ -WebSession $Session -NoProxy:$false).StatusCode
     }
 }
```

CarloToso replied:
I don't know if my idea is correct, but what if once you set -NoProxy on the first request it stays NoProxy=true until you use -Proxy in a subsequent request? (And vice versa)

stevenebutler replied:
I haven't thought about that before, but I have a suspicion it would make it even more complex.

stevenebutler replied:
See latest commit(s) as I have implemented this idea.

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -74,28 +161,95 @@ public class WebRequestSession
-            // build the cookie jar
-            Cookies = new CookieContainer();
+            // Build the cookie jar
+            _cookies = new CookieContainer();
```
If I understand the code correctly you can revert _cookies -> Cookies (and the same for all the others in WebRequestSession())

stevenebutler replied:
You could, but it would go through a longer code path checking for change when we already know it's changed. Is that the preferred option?

CarloToso replied:
I think @iSazonov will answer to this question

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -837,7 +842,11 @@ internal virtual void ValidateParameters()
         {
             // Make sure we have a valid WebRequestSession object to work with
-            WebSession ??= new WebRequestSession();
+            _noWebSession = WebSession is null;
```
```cs
_noWebSession = WebSession is null;
WebSession ??= new WebRequestSession();
```

stevenebutler replied:
Changed this as recommended.

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -909,15 +922,26 @@ internal virtual void PrepareSession()
+                    WebSession.Proxy = webProxy;
+                }
+            }
+            if (MyInvocation.BoundParameters.ContainsKey("SslProtocol"))
```
I don't think we need to check for SslProtocol
```cs
WebSession.SslProtocol = SslProtocol
```
should be ok

stevenebutler replied:
SslProtocol is a struct type so if it's not provided on the command line it will be the default value. That would mean we'd require the user to always supply the SslProtocol on every IWR if they chose a non-default value unless we check for it explicitly being present before deciding whether to use it to 'change' the webrequestsession, which would then cause us to open a new request.

The logic is if the user doesn't supply the parameter at all, but they've provided a Session, we should use whatever was set in the Session, not go back to the default.

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebProxy.cs:
```diff
@@ -18,6 +18,34 @@ internal WebProxy(Uri address)
+            return Equals(obj as WebProxy);
+        }
+
+        public override int GetHashCode()
```
```suggestion
        public override int GetHashCode() => _proxyAddress.GetHashCode() * 17 + _credentials.GetHashCode();
```

stevenebutler replied:
Updated, but was forced to add additional parenthesis by the style checker.

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebProxy.cs:
```diff
@@ -18,6 +18,34 @@ internal WebProxy(Uri address)
+            }
+
+            // _proxyAddress cannot be null as it is set in the constructor            
+            return other._credentials == _credentials
```
```suggestion
            return GetHashCode() == other.GetHashCode() && BypassProxyOnLocal == other.BypassProxyOnLocal;
```

stevenebutler replied:
I would have thought even with comparing hashcode as a fast fail we should still make sure all other properties match for this to be considered equal.

```c#
          return GetHashCode() == other.GetHashCode() 
                && other._credentials == _credentials
                && _proxyAddress.Equals(other._proxyAddress)
                && BypassProxyOnLocal == other.BypassProxyOnLocal;
```

CarloToso replied:
> > Maybe you could do something like this for WebRequestSession.cs (it's just an example I think it can be done better)
> > ```cs
> >         internal int MaxAutomaticRedirections
> >         {
> >             get => _maxAutomaticRedirections;
> >             set => _maxAutomaticRedirections = SetVariable(typeof(int), _maxAutomaticRedirections, value);
> >         }
> > 
> >         internal dynamic SetVariable(Type type, object oldValue, object newValue)
> >         {
> >             dynamic convertedOldValue = Convert.ChangeType(oldValue, type);
> >             dynamic convertedNewValue = Convert.ChangeType(newValue, type);
> >             if (convertedOldValue != convertedNewValue)
> >             {
> >                 _changed = true;
> >                 return convertedNewValue;
> >             }
> >             else
> >             {
> >                 return convertedOldValue;
> >             }
> >         }
> > ```
> 
> I used a couple of generics to do something similar. I'm not a fan of using dynamic since it turns off lots of type checking.
> 
> ```cs
>    private void SetClassVar<T>(ref T oldValue, T newValue) where T : class
>         {
>             if (oldValue == newValue)
>             {
>                 return;
>             }
>             _changed = true;
>             oldValue = newValue;
>         }
> 
>         private void SetStructVar<T>(ref T oldValue, T newValue) where T : struct
>         {
>             if (oldValue.Equals(newValue))
>             {
>                 return;
>             }
>             _changed = true;
>             oldValue = newValue;
>         }
> ```
> 
> Let me know if there's a better way. I wasn't able to use a single generic for both structs and classes but I'm not an expert in this area and someone else may have a better suggestion.

I like your solution 👍

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -13,6 +14,20 @@ namespace Microsoft.PowerShell.Commands
     /// </summary>
     public class WebRequestSession
     {
+        private bool _changed;
```
You could divide using #region; #endregion like in WebRequestPSCmdlet.Common.cs
You could move the getter setters to the top and sort them

stevenebutler replied:
My concern with such changes is it will cause conflicts if others are working on this file too. Maybe once it's closer to getting accepted.

CarloToso replied:
You are right, we can move stuff in a follow up PR after we merge this PR

stevenebutler commented:
> Maybe you could do something like this for WebRequestSession.cs (it's just an example I think it can be done better)
> 
> ```cs
>         internal int MaxAutomaticRedirections
>         {
>             get => _maxAutomaticRedirections;
>             set => _maxAutomaticRedirections = SetVariable(typeof(int), _maxAutomaticRedirections, value);
>         }
> 
>         internal dynamic SetVariable(Type type, object oldValue, object newValue)
>         {
>             dynamic convertedOldValue = Convert.ChangeType(oldValue, type);
>             dynamic convertedNewValue = Convert.ChangeType(newValue, type);
>             if (convertedOldValue != convertedNewValue)
>             {
>                 _changed = true;
>                 return convertedNewValue;
>             }
>             else
>             {
>                 return convertedOldValue;
>             }
>         }
> ```

I used a couple of generics to do something similar. I'm not a fan of using dynamic since it turns off lots of type checking.

```c#
   private void SetClassVar<T>(ref T oldValue, T newValue) where T : class
        {
            if (oldValue == newValue)
            {
                return;
            }
            _changed = true;
            oldValue = newValue;
        }

        private void SetStructVar<T>(ref T oldValue, T newValue) where T : struct
        {
            if (oldValue.Equals(newValue))
            {
                return;
            }
            _changed = true;
            oldValue = newValue;
        }
```
Let me know if there's a better way. I wasn't able to use a single generic for both structs and classes but I'm not an expert in this area and someone else may have a better suggestion.

stevenebutler committed:
Address review comments by @CarloToso

stevenebutler committed:
Make -NoProxy and -Proxy switches flip each other's states

Addresses review comment by @CarloToso

stevenebutler committed:
Address codefactor issues

CarloToso commented:
@stevenebutler you could look into adding some tests to WebCmdlets.Tests.ps1

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -943,57 +972,75 @@ internal virtual void PrepareSession()
+                handler.SslProtocols = (SslProtocols)WebSession.SslProtocol;
+            }
+            // Only dispose the httpClientHandler if we are not trying to use a persistent one
+            HttpClient httpClient = new(handler, _noWebSession && string.IsNullOrEmpty(SessionVariable));
```
Maybe you can replace _noWebSession with a helper method _noWebSession => MyInvocation.BoundParameters.ContainsKey("WebSessionl"))

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -74,39 +111,92 @@ public class WebRequestSession
+        /// <summary>
+        /// Gets a value indicating whether a new handler must be created.
+        /// </summary>
+        internal bool Changed => _changed;
```
If follow my previous comment we have no need this since if we try to change a property already started HttpClient/handler we can catch InvalidOperationException
and set _handler to null. (Of cause we need GetHttpClient/GetHttpClientHandler method in WebSession.)

stevenebutler replied:
See my other comments for the limitations of the approach you propose. It is probably a simpler approach but only if we don't care that any use of httpclienthandler affecting parameters on the command line with a WebSession are going to cause non-persistence.

iSazonov replied:
We can use only `Gets` and check if an argument was changed and if so we recreate the target object.

stevenebutler commented:
You would either have to accept that persistence would not work if the user
supplied any client handler impacting parameters to iwr or you would have
to check every assignment to the handler for logical equivalence before
attempting to set it with this approach.

You could argue that to get persistence they should clean up the script
parameters. I was going for maximum performance with least script change
effort and maximum compatibility with this PR.


On Mon, 20 Feb 2023, 4:10 am Ilya, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs
> <https://github.com/PowerShell/PowerShell/pull/19173#discussion_r1111294075>
> :
>
> > @@ -13,6 +14,20 @@ namespace Microsoft.PowerShell.Commands
>      /// </summary>
>      public class WebRequestSession
>      {
> +        private bool _changed;
> +        private CookieContainer _cookies;
> +        private bool _useDefaultCredentials;
> +        private ICredentials _credentials;
> +        private X509CertificateCollection _certificates;
> +        private IWebProxy _proxy;
> +        private int _maximumRedirection;
> +        private HttpClientHandler _handler;
>
> If we keep HttpClientHandler (and perhaps HttpClient) we could do not keep
> values in WebSession since they are already in HttpClientHandler.
> ------------------------------
>
> In
> src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs
> <https://github.com/PowerShell/PowerShell/pull/19173#discussion_r1111295483>
> :
>
> >              UserAgent = PSUserAgent.UserAgent;
>
> -            Proxy = null;
> -            MaximumRedirection = -1;
> +            _proxy = null;
> +            _maximumRedirection = -1;
> +            _allowAutoRedirect = true;
> +        }
> +
> +        /// <summary>
> +        /// Gets a value indicating whether a new handler must be created.
> +        /// </summary>
> +        internal bool Changed => _changed;
>
> If follow my previous comment we have no need this since if we try to
> change a property already started HttpClient/handler we can catch
> InvalidOperationException
> and set _handler to null. (Of cause we need
> GetHttpClient/GetHttpClientHandler method in WebSession.)
> ------------------------------
>
> In
> src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs
> <https://github.com/PowerShell/PowerShell/pull/19173#discussion_r1111288615>
> :
>
> >
> -            HttpClient httpClient = new(handler);
> +                if (WebSession.SkipCertificateCheck)
> +                {
> +                    handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
> +                    handler.ClientCertificateOptions = ClientCertificateOption.Manual;
> +                }
> +                handler.AllowAutoRedirect = WebSession.AllowAutoRedirect;
> +                if (WebSession.AllowAutoRedirect && WebSession.MaximumRedirection > 0)
> +                {
> +                    handler.MaxAutomaticRedirections = WebSession.MaxAutomaticRedirections;
> +                }
> +                handler.SslProtocols = (SslProtocols)WebSession.SslProtocol;
> +            }
> +            // Only dispose the httpClientHandler if we are not trying to use a persistent one
> +            HttpClient httpClient = new(handler, _noWebSession && string.IsNullOrEmpty(SessionVariable));
>
> Since we don't use HttpClientHandler without HttpClient and vise verse we
> could keep both HttpClient and HttpClientHandler. In the case we have no
> need to have _noWebSession.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PowerShell/PowerShell/pull/19173#pullrequestreview-1304839397>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEADFJP3GSY4TSH73FKQPULWYJO2HANCNFSM6AAAAAAVAEFQDU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

stevenebutler commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -13,6 +14,20 @@ namespace Microsoft.PowerShell.Commands
+        private X509CertificateCollection _certificates;
+        private IWebProxy _proxy;
+        private int _maximumRedirection;
+        private HttpClientHandler _handler;
```
The values in httpclienthandler are processed versions of the ones in the websession. We would still need some way of comparing what has been set with what we are needing to change because even if we set the same value on the HttpClientHandler that we've already set, it will still throw if it's already done a request. The setters all throw immediately if a request has already been issued, they don't do any kind of comparison with the value you're trying to set and the value they already have.

iSazonov replied:
Yes, we need the checks. I say we have no need to duplicate - we can get properties from target object. If any check fails we will recreate target object.

stevenebutler replied:
This would save some memory in the WebRequestSession used to cache the settings by adding considerable additional conditionals to the path to setup the handler in the WebRequestPSCmdlet.Common.cs.

There is not a one-one mapping between the parameters to IWR and what is set on the handler in all cases, so we'd need to either duplicate all that logic in a validation phase, or put it inline in the code that creates it, with the need to then double back and start again once anything fails the validation. The additional checks will increase the number of conditionals in the cmdlet which leads to increased complexity and difficulty to maintain. I don't think this will make the code simpler or easier to understand, but perhaps I'm misunderstanding what you are suggesting.

iSazonov replied:
We have 6 scenario:
1. Use WebSession only internally. (The cmdlet must dispose the WebSession.)
2. The cmdlet issues the WebSession and it is reused as-is
3. The cmdlet issues the WebSession and it is modified by users before reuse.
4. User creates the WebSession and it is reused as-is
5. User creates the WebSession, uses in cmdlet and it is reused as-is
6. User creates the WebSession, uses in cmdlet, modify and reused

We use WebSession as argument store. So I'd expect all logic is in WebSession.

stevenebutler replied:
Perhaps I'm misunderstanding your change proposal, since I agree with your 6 scenarios above and the patch is using the WebSession parameter as an argument store. Apart from the `HttpClientHandler _handler`, all those properties you listed (repeated below) already exist in the `WebRequestSession` class. The patch converts them to non-auto properties so the setter can detect if anyone is altering them through the public APIs and/or IWR parameters. I also added a few other properties that alter the properties of the `HttpRequestHandler` so I can determine if they have also been changed by the user at the time of invocation.

```csharp
        private X509CertificateCollection _certificates;
        private IWebProxy _proxy;
        private int _maximumRedirection;
        private HttpClientHandler _handler;
```

iSazonov replied:
Okay, let's see how your implementation works and make improvements if necessary.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -437,6 +437,11 @@ public virtual string CustomMethod
+        /// <summary>
+        /// True if the web session was created specifically by this request
+        /// </summary>
+        private bool _noWebSession;
```
This is used only once for conditional disposing HttpClientHandler.
I suggest to keep HttpClient in WebSession too since they exist only in couple.
Then _noWebSession can be removed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -894,30 +900,53 @@ internal virtual void PrepareSession()
             }
-
-            if (Proxy is not null)
+            // Proxy and NoProxy parameters are mutually exclusive
+            // If NoProxy is provided, WebSession will turn off the proxy
+            // and if Proxy is provided NoProxy will be turned off
+            if (NoProxy.IsPresent)
```
We have code conventions. Please look docs and existing code. Also please look CI CodeFactor.

Here please and new line after `}` and before the comment.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -894,30 +900,53 @@ internal virtual void PrepareSession()
-                    // UseDefaultCredentials will overwrite the supplied credentials.
-                    webProxy.UseDefaultCredentials = true;
+                    WebProxy webProxy = new(Proxy);
+                    webProxy.BypassProxyOnLocal = false;
```
/cc @CarloToso It seems we don't use `BypassProxyOnLocal`. If you have an interest please open new tracking issue for follow investigations.

CarloToso replied:
`BypassProxyOnLocal==true` can be set if we add a new Proxy to the WebSession and then we use it.
Do you think we should set `BypassProxyOnLocal true or false` directly from WebRequestPSCmdlet (if so I'll open a tracking issue)?

iSazonov replied:
@CarloToso I say we never use a value of the BypassProxyOnLocal property.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -943,57 +972,75 @@ internal virtual void PrepareSession()
-            HttpClientHandler handler = new();
-            handler.CookieContainer = WebSession.Cookies;
-            handler.AutomaticDecompression = DecompressionMethods.All;
+            var handler = WebSession.Handler;
```
I think it should be a method. Since WebSession is a store for (all) parameters we can move all (below) logic in WebSession. This assumes handleRedirect is in WebSession too.
```c#
        internal virtual HttpClient GetHttpClient(bool handleRedirect)
            => WebSession.GetHttpClient();
```

stevenebutler replied:
Have forwarded GetHttpClient() through to WebSession as suggested. Note that to log warnings when the connection changes I needed to pass out a variable indicating the client was recreated.

I have added tests that rely on the LogWarning() message to verify the connection handling is working as expected.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebProxy.cs:
```diff
@@ -18,6 +18,32 @@ internal WebProxy(Uri address)
 
+        public override bool Equals(object obj)
+        {
+            if (obj == null || GetType() != obj.GetType())
```
I guess the type comparison is covered in line 28. 
```suggestion
            if (obj == null)
```

stevenebutler replied:
Type comparison and null check are both covered in the Equals() so I've just forwarded the function directly after casting as WebProxy.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebProxy.cs:
```diff
@@ -18,6 +18,32 @@ internal WebProxy(Uri address)
+            return Equals(obj as WebProxy);
+        }
+
+        public override int GetHashCode() => (_proxyAddress.GetHashCode() * 17) + (_credentials?.GetHashCode() ?? 0);
```
Please use `System.HashCode.Combine()`

stevenebutler replied:
Thanks, have updated as suggested.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -13,6 +14,20 @@ namespace Microsoft.PowerShell.Commands
     /// </summary>
     public class WebRequestSession
     {
+        private bool _changed;
```
We can remove the field and use _httpClient field instead.
If we will create HttpClient in GetHttpClient() method this assumes the HttpClient will be started and can not be changed, if any parameter will be really changed, the fields should be set to null.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -74,39 +111,92 @@ public class WebRequestSession
+        /// <summary>
+        /// Gets or sets the cached HttpClientHandler in the session to support persistent HTTP connections.
+        /// </summary>
+        internal HttpClientHandler Handler
```
I suggest to keep HttpClient and create it by demand with GetHttpClient() method.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -74,39 +111,92 @@ public class WebRequestSession
+            }
+        }
+
+        private void SetClassVar<T>(ref T oldValue, T newValue) where T : class
+        {
+            if (oldValue != newValue)
+            {
+                _changed = true;
+                oldValue = newValue;
+            }
+        }
```
I suggest set to null _httpClient field here.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -74,39 +111,92 @@ public class WebRequestSession
+        }
 
-            Certificates.Add(certificate);
+        private void SetStructVar<T>(ref T oldValue, T newValue) where T : struct
+        {
+            if (!oldValue.Equals(newValue))
+            {
+                _changed = true;
+                oldValue = newValue;
+            }
```
The same.

stevenebutler committed:
Address @iSazanov review comments and suggestions

* Move GetHttpClient body to WebRequestSession
* Make internal properties setter only
* Implement IDisposable for WebRequestSession
* Add tests for persistence
* Fix GetHashCode() / Equals() on WebProxy

stevenebutler committed:
Address codefactor issues

stevenebutler committed:
Address more codefactor issues

stevenebutler committed:
Tidy up comment

stevenebutler commented:
@iSazonov and @CarloToso - I believe I have addressed the review comments as best I can, following @iSazonov 's suggestions for moving GetHttpClient.

I have also added some tests that rely on the LogWarning() message to validate persistence is working as expected. Happy to use a different approach if you know of a better way to do this.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -517,7 +517,7 @@ protected override void ProcessRecord()
                 bool handleRedirect = keepAuthorizationOnRedirect || AllowInsecureRedirect || PreserveHttpMethodOnRedirect;
 
-                using (HttpClient client = GetHttpClient(handleRedirect))
+                var client = GetHttpClient(handleRedirect);
```
Please replace `var` with real type for readability.

stevenebutler replied:
Done

CarloToso replied:
You could also remove { } and one level of indentation

stevenebutler replied:
I deliberately backed out that change as I thought it would make later merges more difficult and also inflate the size of the PR and make it harder to review. I can still do it if you think it is best.

CarloToso replied:
Don't worry I'll do it in a follow up PR

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -683,6 +683,23 @@ protected override void ProcessRecord()
+            base.EndProcessing();
+            if (!IsPersistentSession())
+            {
+                WebSession?.Dispose();
```
Cmdlet class has Dispose(). So we could move this in that overload.

stevenebutler replied:
I can only see an internal InternalDispose() method which is not virtual. I'm unsure what I should do here to not cause a leak. I was fairly certain EndProcessing was marked as the cleanup method for CmdLets which is why I implemented here.

Could you provide more info on how I would implement the Dispose() - should I make the cmdlet implement IDisposable or override the InternalDispose() method?

iSazonov replied:
Sorry, I suggest to make the cmdlet IDisposable.

stevenebutler replied:
ok, have done this.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -683,6 +683,23 @@ protected override void ProcessRecord()
+            }
+        }
+
+        private bool IsPersistentSession()
+        {
+            return MyInvocation.BoundParameters.ContainsKey("WebSession") || !string.IsNullOrEmpty(SessionVariable);
+        }
```
Please use the same detection for SessionVariable. Also we could use short definition with `=>`.

stevenebutler replied:
Done. Have also moved into helper functions region

stevenebutler committed:
Merge branch 'PowerShell:master' into 12764-support-persistent-http-connections-in-cmdlets

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -895,29 +912,53 @@ internal virtual void PrepareSession()
                 }
-
-                WebSession.Proxy = webProxy;
+            }
+            if (MyInvocation.BoundParameters.ContainsKey("SslProtocol"))
```
```suggestion
            }

            if (MyInvocation.BoundParameters.ContainsKey("SslProtocol"))
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -943,62 +984,16 @@ internal virtual void PrepareSession()
 
-            // Set the credentials used by this request
-            if (WebSession.UseDefaultCredentials)
+            var client = WebSession.GetHttpClient(handleRedirect, TimeoutSec, out var newClient);
```
Please replace `var` with real types.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -943,62 +984,16 @@ internal virtual void PrepareSession()
             {
-                // The UseDefaultCredentials flag overrides other supplied credentials
-                handler.UseDefaultCredentials = true;
+                WriteWarning("WebSession properties changed: new Http connection will be required");
```
It seems the condition is wrong.

stevenebutler replied:
I don't know what you're referring to here. Could you provide more detail? If it is related to the tests failing, I have identified that as being caused by changing the MaxAutomaticRedirections property on the WebSession while processing the CmdLet, which I had not realised was occurring. This caused the WebRequestSession to Dispose of the in-use client.

This should be resolved in my latest commit, which includes the changes by @CarloToso from #19190 as well as additional changes suggested to remove the property entirely from the WebRequestSession.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -943,62 +984,16 @@ internal virtual void PrepareSession()
-            httpClient.Timeout = TimeoutSec is 0 ? TimeSpan.FromMilliseconds(Timeout.Infinite) : new TimeSpan(0, 0, TimeoutSec);
-
-            return httpClient;
+            return client;
```
```suggestion

            return client;
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebProxy.cs:
```diff
@@ -18,6 +18,24 @@ internal WebProxy(Uri address)
+            }
+
+            // _proxyAddress cannot be null as it is set in the constructor            
+            return GetHashCode() == other.GetHashCode()
+                && other._credentials == _credentials
+                && _proxyAddress.Equals(other._proxyAddress)
+                && BypassProxyOnLocal == other.BypassProxyOnLocal;
```
Do we really need calculate and compare hashes?

stevenebutler replied:
I did not have that in there initially. I added in response to someone's suggestions. I can remove it if you like.

CarloToso replied:
Remove it 👍

stevenebutler replied:
Done

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -57,12 +75,22 @@ public class WebRequestSession
-        public IWebProxy Proxy { get; set; }
+        public IWebProxy Proxy
+        {
+            get => _proxy; set
```
Please move `set` to new line.

stevenebutler replied:
Done.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -105,8 +154,142 @@ public WebRequestSession()
+
+        private void ResetClient()
+        {
+            _client?.Dispose();
```
Do we reset client on every changed parameter? If so it is not correct.

stevenebutler replied:
We reset client on every parameter that would cause it to throw an exception if we tried to update it.

This includes anything that would impact the HttpClientHandler or the client itself (you can't change the timeout on the client once it has issued a request).

Note that there is a problem with the latest refactor that is breaking some tests. I am trying to track down what is wrong at the moment.

stevenebutler replied:
Changes in https://github.com/PowerShell/PowerShell/pull/19190 resolve the test failures in my PR.

stevenebutler committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs

Co-authored-by: Ilya <darpa@yandex.ru>

stevenebutler committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs

Co-authored-by: Ilya <darpa@yandex.ru>

CarloToso commented:
@stevenebutler After we merge #19190 (hopefully soon), I think we can remove `_maxAutomaticRedirections` and just use `_maximumRedirection`

stevenebutler commented:
> @stevenebutler After we merge #19190 (hopefully soon), I think we can remove `_maxAutomaticRedirections` and just use `_maximumRedirection`

@CarloToso I need the changes in #19190 for the unit tests to pass. Using the WebRequestSession property to decrement causes it to unexpectedly Dispose of the client in the middle of using it for the redirects and was failing the tests. With your changes integrated into my PR (I applied manually to my branch), the tests pass again.

Are you talking about removing MaxAutomaticRedirections completely from the WebRequestSession class or some other simplification? I'm not sure what you mean by removing _maxAutomaticRedirections.

CarloToso commented:
@stevenebutler Yes, I was talking about removing `MaxAutomaticRedirections` completely from the WebRequestSession. Once we stop decrementing `MaximumRedirection` they should always have the same value so we don't need both.

stevenebutler committed:
Address review comments

stevenebutler committed:
Remove unnecessary MaximumRedirection property

stevenebutler committed:
Simplify logic around whether httpClient was disposed

stevenebutler commented:
> @stevenebutler Yes, I was talking about removing `MaxAutomaticRedirections` completely from the WebRequestSession. Once we stop decrementing `MaximumRedirection` they should always have the same value so we don't need both.

@CarloToso I have applied your changes from that PR manually as well as the suggestion to remove the property from `WebRequestSession`. This appears to have fixed the test failures on my PC, but I'll wait to see what happens with the CI build.

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -943,62 +987,18 @@ internal virtual void PrepareSession()
-                handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
-                handler.ClientCertificateOptions = ClientCertificateOption.Manual;
-            }
 
```
```suggestion
```

stevenebutler commented:
@CarlToso I'm not sure how this test relates to my changes, but it failed in CI. Does it seem likely my PR has caused this failure?

It seems to be a test that is reading windows events. Would the additional LogWarning() be the cause of this or is it likely just a test timing/flakiness issue?

![image](https://user-images.githubusercontent.com/16790181/221356629-467704b7-22b4-492a-9b61-865c822ca377.png)

Test that failed is on line 94 of test\powershell\engine\Basic\GroupPolicySettings.Tests.ps1:

```powershell
 Wait-UntilTrue -sb { Get-WinEvent -FilterHashtable @{ ProviderName="PowerShellCore"; Id = 4103 } -MaxEvents 5 |
                    Where-Object {$_.Message.Contains($RareCommand)} } -TimeoutInMilliseconds (5*1000) -IntervalInMilliseconds 100 |
                        Should -BeTrue
```

CarloToso commented:
@stevenebutler I have already seen this failure other times in the last month, it's unrelated to this PR

stevenebutler committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs

Co-authored-by: CarloToso <105941898+CarloToso@users.noreply.github.com>

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -943,62 +987,17 @@ internal virtual void PrepareSession()
-                handler.AllowAutoRedirect = false;
-            }
-            else if (WebSession.MaximumRedirection > 0)
+            HttpClient client = WebSession.GetHttpClient(handleRedirect, TimeoutSec, out bool clientWasReset);
```
Can we evaluate `handleRedirect`internally in WebSession?
Can we save TimeoutSec in WebSession too?

stevenebutler replied:
We can evaluate the TimeoutSec in WebSession. I have made this change.

The handleRedirect flag is not so obvious. It is controlled by a lot of different factors, including a Http Header being present in the current request that we cannot handle in WebSession without adding a lot of duplicated properties too. I think this would likely be more complicated and confusing than the current approach.

Unfortunately, turning on or off automatic (.NET) redirection support means we need to create a new HttpClientHandler and we lose the persistent connections. A possible future PR would be to have PowerShell always do the redirect handling itself - in that way it wouldn't matter what redirect settings were tweaked between invocations on a session. It could use the same handler for requests (so long as other properties are not changed).

@CarloToso & @iSazonov  - what do you think of this approach?

CarloToso replied:
I think it should work but it might be slower than the current implementation (currently PowerShell [with handleRedirect==true] is slower than .NET when handling redirects). I don't think we should do it, at least not at the moment. Instead if you want you could try to work on solving this 3 issues #12249, #16122, #16145 (they are closely related to eachother).

stevenebutler replied:
I think it's okay to break the persistence occasionally for unusual scenarios where redirect handling is differing between requests. It will still be considerably better than it was before with no persistence for any connections for most users.

I may have a look at those issues, but not while this one is still in progress.

CarloToso replied:
Yes, my suggestion was for a future PR.

stevenebutler committed:
Implements IDisposable to dispose WebSession

As per review request from @iSazonov, WebRequestPSCmdlet implements
IDisposable and cleans up non-persistent WebSession objects in the
Dispose method.

stevenebutler committed:
Move TimeoutSec handling to WebRequestSession

Responding to review comments from @iSazonov

CarloToso commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,161 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+                    $null = Invoke-WebRequest -Uri $uri -SkipCertificateCheck
+                }
+            }
+            $persistence | Should -belessthan $nonpersistence
```
```suggestion
            $persistence | Should -BeLessThan $nonpersistence
```

CarloToso commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,161 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+                $caught = $true
+            }
+            $Warnings | Should -be $connWarning
+            $caught | should -BeTrue
```
```suggestion
            $caught | Should -BeTrue
```

CarloToso commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,161 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+            } catch {
+                $caught = $true
+            }
+            $Warnings | Should -be $connWarning
```
```suggestion
            $Warnings | Should -Be $connWarning
```

CarloToso commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,161 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+
+        It 'Connection persistence maintained' {
+            $Uri = Get-WebListenerUrl
+            $null = Invoke-WebRequest -Uri $uri -SessionVariable Session -wv warnings
```
```suggestion
            $null = Invoke-WebRequest -Uri $Uri -SessionVariable Session -WarningVariable Warnings
```
Could you please, always use variables with the same capitalization in the tests

CarloToso commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,161 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+        }
+
+        It 'Connection persistence outperforms non-persistence' {
+            $Uri = Get-WebListenerUrl -Https
```
```suggestion
            $uri = Get-WebListenerUrl -Https
```

stevenebutler replied:
I have addressed the naming issues in the latest commit.

CarloToso replied:
I think you can remove the W.I.P label

stevenebutler committed:
Clean up variable naming in persistence test cases

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -683,6 +695,37 @@ protected override void ProcessRecord()
+        {
+            if (!_disposedValue)
+            {
+                if (disposing)
+                {
+                    if (!IsPersistentSession())
+                    {
+                        WebSession?.Dispose();
+                        WebSession = null;
+                    }
+                }
```
Please combine the two `if`-s to one.

stevenebutler replied:
Done in latest PR

stevenebutler committed:
Combine if expressions in dispose

As per review request by @iSazonov

stevenebutler committed:
Remove indentation

stevenebutler committed:
Merge indentation change from master

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -107,6 +107,11 @@ public abstract class WebRequestPSCmdlet : PSCmdlet
+        /// <summary>
+        /// Maximum number of Redirects to follow, caches WebSession.MaximumRedirection.
+        /// </summary>
+        internal int _maximumRedirection;
```
Please make the comment better.

stevenebutler replied:
I have changed the comment, but not sure if it qualifies as better. Please use more actionable feedback or my changes may be misguided and make things worse, not better.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -127,6 +132,11 @@ public abstract class WebRequestPSCmdlet : PSCmdlet
+        /// <summary>
+        /// The Dispose() method has already been called to cleanup Disposable fields.
+        /// </summary>
+        private bool _disposedValue = false;
```
Please make the comment better.

stevenebutler replied:
I have changed the comment, but not sure if it qualifies as better.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -682,6 +693,34 @@ protected override void ProcessRecord()
+        /// <summary>
+        /// Dispose the CmdLet, cleaning up resources if required. The WebSession must be disposed
+        /// if it is not being used as part of a persistent session.
+        /// </summary> 
```
Please make the comment better.

stevenebutler replied:
I have changed the comment, but not sure if it qualifies as better.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -894,29 +933,54 @@ internal virtual void PrepareSession()
+                    {
+                        // If both ProxyCredential and ProxyUseDefaultCredentials are passed,
+                        // UseDefaultCredentials will overwrite the supplied credentials.
+                        webProxy.UseDefaultCredentials = true;
```
The comment is not correct since the code is in `else` branch.

stevenebutler replied:
Yes, you are right. However this is what was there before, which makes me wonder if it's a bug, or just a bad comment?

Should we fix the comment or the code?

iSazonov replied:
@CarloToso Could you please look this?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -894,29 +933,54 @@ internal virtual void PrepareSession()
+                        // UseDefaultCredentials will overwrite the supplied credentials.
+                        webProxy.UseDefaultCredentials = true;
+                    }
+                    // We don't want to update the WebSession unless the proxies are different
```
```suggestion

                    // We don't want to update the WebSession unless the proxies are different
```

stevenebutler replied:
changed

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -894,29 +933,54 @@ internal virtual void PrepareSession()
+                // SslProtocol parameter is a struct and we only want to switch back to the default
+                // if it is explicitly provided on the command line. Otherwise we keep the value in
+                // the WebSession from any previous invocation.
+                WebSession.SslProtocol = SslProtocol;
```
Struct?
Please make the comment better.

stevenebutler replied:
Changed to "value type". Also applied nameof(SslProtocol) to the test on 969 as per your other comments..

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -941,63 +1005,20 @@ internal virtual void PrepareSession()
             {
-                handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
-                handler.ClientCertificateOptions = ClientCertificateOption.Manual;
+                WriteWarning("WebSession properties changed: new Http connection will be required");
```
I think it should be WriteVerbose. And please move the message to Resx file.

stevenebutler replied:
It was a warning because it indicates to the caller they may be suffering performance issue due to the way they're using the WebSession. If it's set to Verbose they'll never know they're using it in a way that can't support persistence.

I can change it, but please consider the above.

stevenebutler replied:
I have changed to a warning with latest commit.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1455,6 +1476,8 @@ private void ProcessAuthentication()
         }
 
+        private bool IsPersistentSession() => MyInvocation.BoundParameters.ContainsKey("WebSession") || MyInvocation.BoundParameters.ContainsKey("SessionVariable");
+
```
Can we use `nameof` here?

stevenebutler replied:
Yes, good idea.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -105,8 +158,134 @@ public WebRequestSession()
+            }
+        }
+
+        internal virtual HttpClient GetHttpClient(bool handleRedirect, out bool clientWasReset)
```
Why virtual?

stevenebutler replied:
Accidentally copied from the original location. Removed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -105,8 +158,134 @@ public WebRequestSession()
+        internal virtual HttpClient GetHttpClient(bool handleRedirect, out bool clientWasReset)
+        {
+            // This indicates GetResponse will handle redirects.
+            SetStructVar(ref _allowAutoRedirect, !(handleRedirect || MaximumRedirection == 0));
```
Please make the comment better.

stevenebutler replied:
Comment updated - this comment makes less sense now the code has been moved from the webrequest cmdlet base class. handleRedirect is also reverse of what it actually does, so I've renamed that here to doNotRedirect and added documentation on the class.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,174 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+            $null = Invoke-WebRequest -Uri $uri -SessionVariable Session -WarningVariable warnings -SkipCertificateCheck
+            $iters = 5
+            $persistence = Measure-Command { 1 .. $iters | ForEach-Object {
+                    $null = Invoke-WebRequest -Uri $uri -WebSession $Session -SkipCertificateCheck
```
Indentation should be 4. Below too.

stevenebutler replied:
Have moved the 1 .. $iters down a line so indentation is correct

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,174 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+            $warnings | Should -BeNullOrEmpty
+        }
+
+        It 'proxy can stay off between tests' {
```
```suggestion
        It 'Proxy can stay off between tests' {
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,174 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+            $warnings | Should -BeNullOrEmpty
+        }
+
+        It 'proxy can turn on and off between tests' {
```
```suggestion
        It 'Proxy can turn on and off between tests' {
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,174 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+    }
 }
 
 Describe "Invoke-RestMethod tests" -Tags "Feature", "RequireAdminOnWindows" {
```
We have to duplicate the tests for Invoke-RestMethod cmdlet.

stevenebutler replied:
Let's get it right for IWR then I'll look at duplicating.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2193,174 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+            $null = Invoke-WebRequest -Uri $uri -WebSession $Session -WarningVariable warnings
+            $warnings | Should -BeNullOrEmpty
+
+            $null = Invoke-WebRequest -Uri $uri -WebSession $Session -WarningVariable warnings
+            $warnings | Should -BeNullOrEmpty
+
+            $null = Invoke-WebRequest -Uri $uri -WebSession $Session -WarningVariable warnings
+            $warnings | Should -BeNullOrEmpty
```
Why duplicate?

stevenebutler replied:
Mainly to prove it stays persistent for multiple requests. Might be better written as a foreach loop.

stevenebutler committed:
Apply suggestions from code review

Co-authored-by: Ilya <darpa@yandex.ru>

stevenebutler committed:
Address review comments by @iSazonov

Comment improvements
Changed reconnection warning to verbose message
Moved verbose message text to resource
Updated tests

stevenebutler committed:
codefactor fix

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -105,8 +158,141 @@ public WebRequestSession()
+        internal HttpClient GetHttpClient(bool doNotRedirect, out bool clientWasReset)
+        {
+            // Do not auto redirect if the the caller does not want it, or maximum redirections is 0
+            SetStructVar(ref _allowAutoRedirect, !(doNotRedirect || MaximumRedirection == 0));
```
```suggestion
            SetStructVar(ref _allowAutoRedirect, !(doNotHandleRedirect || MaximumRedirection is 0));
```

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -105,8 +158,141 @@ public WebRequestSession()
+        /// Gets an existing or creates a new HttpClient for this WebRequest session if none currently exists (either because it was never
+        /// created, or because changes to the WebSession properties required the existing HttpClient to be disposed).
+        /// </summary>
+        /// <param name="doNotRedirect">True if the caller does not want the HttpClient to ever handle redirections automatically.</param>
```
```suggestion
        /// <param name="doNotRedirect">True if the caller does not want the HttpClient to ever handle redirections automatically.</param>
```
```suggestion
        /// <param name="doNotRedirect">True if the caller does not want the HttpClient to ever handle redirections automatically.</param>
```
```suggestion
        /// <param name="doNotHandleRedirect">True if the caller does not want the HttpClient to ever handle redirections automatically.</param>
```

stevenebutler committed:
Make parameter name clearer as per review suggestion

stevenebutler commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -165,10 +165,15 @@ internal void AddCertificate(X509Certificate certificate)
-            // This indicates GetResponse will handle redirects.
-            SetStructVar(ref _allowAutoRedirect, !(handleRedirect || MaximumRedirection == 0));
+            // Do not auto redirect if the the caller does not want it, or maximum redirections is 0
+            SetStructVar(ref _allowAutoRedirect, !(doNotRedirect || MaximumRedirection == 0));
```
It is handleRedirect from the perspective of cmdlet, but do not from the perspective of the websession's httpclient. I'm not sure what else to call it. Have made the change you suggested to doNotHandleRedirect in the WebSession.

CarloToso replied:
I think `doNotHandleRedirect` is better but it's still not the best (sorry I deleted the suggestion, I was thinking of a more fitting name)

stevenebutler replied:
How about `supressHttpClientRedirects` ?

Also, could you take a look at iSazanov's comment regarding the comment around UseDefaultCredentials. It looks wrong based on the code, but I'm not sure if the code or comment is ultimately correct.

```csharp
                        // If both ProxyCredential and ProxyUseDefaultCredentials are passed,
                        // UseDefaultCredentials will overwrite the supplied credentials.
                        webProxy.UseDefaultCredentials = true;
```

CarloToso replied:
I gave it a quick look and it seems the code could be wrong, not only there but every time we use `UseDefaultCredentials`, when I have some time I will open a tracking issue to investigate.

`supressHttpClientRedirects` might work, what do you think @iSazonov?

iSazonov replied:
Maybe more short `disableRedirect`.

stevenebutler committed:
Rename function parameter to suppressHttpClientRedirects

iSazonov commented:
@stevenebutler The PR contains too many commits and comments. I suggest to close the PR and open new one. Also maybe squash commits - it seems the history is not so useful.

stevenebutler commented:
I am closing this PR in favour of #19249
